File: src/94.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> reordered;
        if (root == NULL) return reordered;
        stack<TreeNode*> nodeStack;
        nodeStack.push(root);
        while (!nodeStack.empty()) {
            TreeNode* node = nodeStack.top();
            if (node->right != NULL) {
                nodeStack.pop();
                nodeStack.push(node->right);
                node->right = NULL;
                nodeStack.push(node);
            }
            if (node->left != NULL) {
                nodeStack.push(node->left);
                node->left = NULL;
                continue;
            }
            nodeStack.pop();
            reordered.push_back(node->val);
        }
        return reordered;
    }
};

int main() {
    Solution solution;
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    auto ans = solution.inorderTraversal(root);
    for (auto it = ans.begin(); it != ans.end(); it++) {
        cout << *it << " ";
    }
    return 0;
}

File: pysrc/94.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    UNCHARTED = 0
    VISITED = 1

    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        stack = [(self.UNCHARTED, root)]
        traversal = []
        while len(stack) != 0:
            state, node = stack.pop()
            if state == self.UNCHARTED:
                if node.right != None:
                    stack.append((self.UNCHARTED, node.right))
                stack.append((self.VISITED, node))
                if node.left != None:
                    stack.append((self.UNCHARTED, node.left))
            elif state == self.VISITED:
                traversal.append(node.val)
        return traversal


if __name__ == '__main__':
    solution = Solution()
    root = TreeNode(1)
    root.left = TreeNode(4)
    root.right = TreeNode(2)
    root.right.left = TreeNode(3)
    print(solution.inorderTraversal(root))
